File: code/packages/ruby/mini_sqlite/lib/coding_adventures/mini_sqlite/binding.rb
# frozen_string_literal: true

module CodingAdventures
  module MiniSqlite
    module Binding
      module_function

      def substitute(sql, params = [])
        index = 0
        out = +""
        i = 0
        while i < sql.length
          ch = sql[i]
          nxt = sql[i + 1]
          if ch == "'" || ch == '"'
            literal, i = read_quoted(sql, i, ch)
            out << literal
            next
          end
          if ch == "-" && nxt == "-"
            j = sql.index("\n", i) || sql.length
            out << sql[i...j]
            i = j
            next
          end
          if ch == "/" && nxt == "*"
            j = sql.index("*/", i + 2)
            j = j ? j + 2 : sql.length
            out << sql[i...j]
            i = j
            next
          end
          if ch == "?"
            raise ProgrammingError, "not enough parameters for SQL statement" if index >= params.length

            out << to_sql_literal(params[index])
            index += 1
            i += 1
            next
          end
          out << ch
          i += 1
        end
        raise ProgrammingError, "too many parameters for SQL statement" unless index == params.length

        out
      end

      def read_quoted(sql, start, quote)
        i = start + 1
        while i < sql.length
          if sql[i] == quote
            if sql[i + 1] == quote
              i += 2
              next
            end
            return [sql[start..i], i + 1]
          end
          i += 1
        end
        [sql[start..], sql.length]
      end

      def to_sql_literal(value)
        case value
        when nil then "NULL"
        when true then "TRUE"
        when false then "FALSE"
        when Integer, Float
          raise ProgrammingError, "non-finite numeric parameter is not supported" unless value.finite?

          value.to_s
        when String
          "'#{value.gsub("'", "''")}'"
        else
          raise ProgrammingError, "unsupported parameter type: #{value.class}"
        end
      end
    end
  end
end

File: code/packages/elixir/mini_sqlite/lib/coding_adventures/mini_sqlite/binding.ex
defmodule CodingAdventures.MiniSqlite.Binding do
  @moduledoc false

  alias CodingAdventures.MiniSqlite.Errors.ProgrammingError

  def bind(sql, params) when is_binary(sql) and is_list(params) do
    scan(sql, params, 0, [])
  end

  defp scan(sql, params, pos, acc) when pos >= byte_size(sql) do
    case params do
      [] ->
        {:ok, acc |> Enum.reverse() |> IO.iodata_to_binary()}

      _ ->
        {:error, %ProgrammingError{message: "too many parameters for SQL statement"}}
    end
  end

  defp scan(sql, params, pos, acc) do
    ch = binary_part(sql, pos, 1)

    cond do
      ch in ["'", "\""] ->
        next = read_quoted(sql, pos, ch)
        scan(sql, params, next, [binary_part(sql, pos, next - pos) | acc])

      ch == "-" and pos + 1 < byte_size(sql) and binary_part(sql, pos + 1, 1) == "-" ->
        next = read_line_comment(sql, pos)
        scan(sql, params, next, [binary_part(sql, pos, next - pos) | acc])

      ch == "/" and pos + 1 < byte_size(sql) and binary_part(sql, pos + 1, 1) == "*" ->
        next = read_block_comment(sql, pos)
        scan(sql, params, next, [binary_part(sql, pos, next - pos) | acc])

      ch == "?" ->
        case params do
          [] ->
            {:error, %ProgrammingError{message: "not enough parameters for SQL statement"}}

          [param | rest] ->
            scan(sql, rest, pos + 1, [to_sql_literal(param) | acc])
        end

      true ->
        scan(sql, params, pos + 1, [ch | acc])
    end
  end

  defp read_quoted(sql, pos, quote) do
    do_read_quoted(sql, pos + 1, quote)
  end

  defp do_read_quoted(sql, pos, _quote) when pos >= byte_size(sql), do: byte_size(sql)

  defp do_read_quoted(sql, pos, quote) do
    ch = binary_part(sql, pos, 1)

    cond do
      ch == "\\" ->
        do_read_quoted(sql, min(pos + 2, byte_size(sql)), quote)

      ch == quote ->
        pos + 1

      true ->
        do_read_quoted(sql, pos + 1, quote)
    end
  end

  defp read_line_comment(sql, pos) do
    do_read_line_comment(sql, pos + 2)
  end

  defp do_read_line_comment(sql, pos) when pos >= byte_size(sql), do: byte_size(sql)

  defp do_read_line_comment(sql, pos) do
    if binary_part(sql, pos, 1) == "\n" do
      pos
    else
      do_read_line_comment(sql, pos + 1)
    end
  end

  defp read_block_comment(sql, pos) do
    do_read_block_comment(sql, pos + 2)
  end

  defp do_read_block_comment(sql, pos) when pos + 1 >= byte_size(sql), do: byte_size(sql)

  defp do_read_block_comment(sql, pos) do
    if binary_part(sql, pos, 2) == "*/" do
      pos + 2
    else
      do_read_block_comment(sql, pos + 1)
    end
  end

  defp to_sql_literal(nil), do: "NULL"
  defp to_sql_literal(true), do: "TRUE"
  defp to_sql_literal(false), do: "FALSE"
  defp to_sql_literal(value) when is_integer(value), do: Integer.to_string(value)
  defp to_sql_literal(value) when is_float(value), do: Float.to_string(value)

  defp to_sql_literal(value) when is_binary(value) do
    value
    |> String.replace("\\", "\\\\")
    |> String.replace("'", "\\'")
    |> String.replace("\n", "\\n")
    |> String.replace("\t", "\\t")
    |> then(&"'#{&1}'")
  end
end
